### PR TITLE
Document (and tweak) panic behavior

### PR DESCRIPTION
**Pull request by notriddle**
Part of #31: C-DOC-PANICS

### CODE REVIEW

**Comment by lnicola** [comment]
looks good, i think. i'd still like to push two commits here, to remove 'n unwr'p 'nd to 'dd some document'tion 'bout p'nics, but i'm h'ving some computer trouble right now :smile:

**Comment by lnicola** [comment]
or if you don't mind doing the work, it's the two lines here https://github.com/notriddle/ammonia/blob/master/src/lib.rs#L451 where 'n unwr'p could be 'voided.

'lso here https://github.com/notriddle/ammonia/blob/master/src/lib.rs#L666, i don't think the write to vec c'n f'il & html5ever is only utf8, i think, so comment could be 'dded.

**Comment by notriddle** [comment]
I hope you can get a good replacement keyboard. If there's any way I can help you get one, just let me know (PM me on the forum or something).

**Comment by lnicola** [comment]
> If there's any way I can help you get a replacement keyboard, just let me know (PM me on the forum or IRC or something). ✋

Thanks. It was a bad USB port, though this isn't my main computer. I'll still need to wait for a replacement charger for my laptop.

**Comment by notriddle** [comment]
Why not use `expect()`?

**Comment by lnicola** [comment]
I suppose we could also extract the consistency checks outside of `clean_dom`, like some other builders do, but I'm not sure it's worth complicating the API.

**Comment by notriddle** [comment]
I don't want to complicate the API, especially since most of these settings will be hard-coded into the app anyway.

File: src/lib.rs [hunk 1]
```diff
@@ -92,6 +92,82 @@ pub fn clean(src: &str) -> String {
...
+///
+///    ```should_panic
+///    #[macro_use]
+///    extern crate maplit;
+///    # extern crate ammonia;
+///    # fn main() {
+///    use ammonia::Builder;
+///    Builder::default()
+///        .generic_attributes(hashset!["rel"])
+///         .clean("");
```
Comment (lnicola) [inline src/lib.rs]: Extra space here

File: src/lib.rs [hunk 2]
```diff
@@ -92,6 +92,82 @@ pub fn clean(src: &str) -> String {
...
+///    ```should_panic
+///    #[macro_use]
+///    extern crate maplit;
+///    # extern crate ammonia;
+///    # fn main() {
+///    use ammonia::Builder;
+///    Builder::default()
+///        .generic_attributes(hashset!["class"])
+///        .allowed_classes(hashmap!["span" => hashset!["hidden"]])
+///         .clean("");
```
Comment (lnicola) [inline src/lib.rs]: Same

File: src/lib.rs [hunk 3]
```diff
@@ -493,7 +577,12 @@ impl<'a> Builder<'a> {
                             .map(|ta| ta.contains(&*attr.name.local)) ==
                             Some(true);
                     if !whitelisted {
-                        false
+                        // If the class attribute is not whitelisted,
+                        // but there is a whitelisted set of allowed_classes,
+                        // do not strip out the class attribute.
+                        // Banned classes will be filtered later.
+                        &*attr.name.local == "class" &&
+                          self.allowed_classes.get(&*name.local).is_some()
```
Comment (lnicola) [inline src/lib.rs]: Maybe `contains_key()` instead. Too bad that the lookup will be performed twice.

**Comment by lnicola** [comment]
Just a couple of nits, r+ otherwise.

**Comment by notriddle** [comment]
bors r= @lnicola, @notriddle

**Pull request merged**